test/parallel/test-promises-unhandled-rejections.js

```diff
@@ -0,0 +1,591 @@
+  });
+});
+
+asyncTest(function(done) {
+  var e = new Error();
+  onUnhandledFail(done);
+  Promise.resolve().then(function() {
+    throw e;
+  }).then(assert.fail, function(reason) {
+    assert.strictEqual(e, reason);
+  });
+});
+
+asyncTest(function(done) {
+  var e = new Error();
+  onUnhandledFail(done);
+  Promise.resolve().then(function() {
+    return Promise.reject(e);
+  }).then(assert.fail, function(reason) {
+    assert.strictEqual(e, reason);
+  });
+});
+
+asyncTest(function(done) {
+  var e = new Error();
+  var _promise;
+  onUnhandledSucceed(done, function(reason, promise) {
+    assert.strictEqual(e, reason);
+    assert.strictEqual(_promise, promise);
+  });
+  _promise = Promise.resolve().then(function() {
+    return new Promise(function(_, reject) {
+      setTimeout(function() {
+        reject(e);
+      }, 1);
+    });
+  });
+});
+
+asyncTest(function(done) {
+  var e = new Error();
+  var _promise;
+  onUnhandledSucceed(done, function(reason, promise) {
+    assert.strictEqual(e, reason);
+    assert.strictEqual(_promise, promise);
+  });
+  _promise = Promise.resolve().then(function() {
+    throw e;
+  });
+});
+
+asyncTest(function(done) {
+  var e = new Error();
+  var _promise;
+  onUnhandledSucceed(done, function(reason, promise) {
+    assert.strictEqual(e, reason);
+    assert.strictEqual(_promise, promise);
+  });
+  _promise = Promise.resolve().then(function() {
+    return Promise.reject(e);
+  });
+});
+
+// Combinations with Promise.all
+asyncTest(function(done) {
+  var e = new Error();
+  onUnhandledFail(done);
+  Promise.all([Promise.reject(e)]).then(assert.fail, function() {});
+});
+
+asyncTest(function(done) {
+  var e = new Error();
+  onUnhandledFail(done);
+  var p = new Promise(function(_, reject) {
+    process.nextTick(function() {
+      reject(e);
+    });
+  });
+  p = Promise.all([p]);
+  process.nextTick(function() {
+    p.then(assert.fail, function() {});
+  });
+});
+
+asyncTest(function(done) {
+  var e = new Error();
+  onUnhandledSucceed(done, function(reason, promise) {
+    assert.strictEqual(e, reason);
+    assert.strictEqual(p, promise);
+  });
+  var p = Promise.all([Promise.reject(e)]);
+});
+
+asyncTest(function(done) {
+  clean();
+  var unhandledPromises = [];
+  var e = new Error();
+  process.on('unhandledRejection', function(reason, promise) {
+    assert.strictEqual(e, reason);
+    unhandledPromises.push(promise);
+  });
+  process.on('rejectionHandled', function(promise) {
+    assert.strictEqual(1, unhandledPromises.length);
+    assert.strictEqual(unhandledPromises[0], promise);
+    assert.strictEqual(thePromise, promise);
+    done();
+  });
+
+  var thePromise = new Promise(function() {
+    throw e;
+  });
+  setTimeout(function() {
+    thePromise.then(assert.fail, function(reason) {
+      assert.strictEqual(e, reason);
+    });
+  }, 10);
+});
+
+
+asyncTest(function(done) {
+  var e = new Error();
+  onUnhandledFail(done);
+
+  setImmediate(function() {
```

Comment:
Because everything behaved differently once setImmediate was triggered vs. when it wasn't, when we used to use setImmediate to do scheduling. There should be no harm in having them even now though.